### ISSUE TITLE
total projects doesn't match projects list

### ISSUE THREAD

**Issue by nsheff** [pepkit/pephub#149]
![image](https://github.com/pepkit/pephub/assets/5271009/4d63b5b4-b6bd-4b21-b1e1-1b0c306576bd)

this says total projects: 22, then it says no projects found.

**Comment by nsheff** [comment]
Also, not sure if that's the same issue or something different, but it turns out if you visit it with a lowercase k, it shows the projects.

So, it appears that pephub is case sensitive. Should it be? 

Is that a separate issue?

**Comment by nleroy917** [comment]
GitHub is case-**insensitive**. `github.com/nleroy917` yields the same as `github.com/NLEROY917`. Dockerhub, however appears case-**sensitive**. `hub.docker.com/r/databio/pephub` **does not** yield the same as `hub.docker.com/r/DATABIO/pephub`. 

I think I am more partial to pephub being case-**insensitive**.

**Comment by nleroy917** [comment]
Via @nsheff 
> github is case-insensitive, and pephub namespaces are using github namespaces, so it should follow the same rules.

That's a strong enough argument for me.

**Comment by nleroy917** [comment]
Likely solved by bringing in the namespace (and project) as lowercase:

```ts
// Project.tsx, Namespace.tsx
  let { namespace, project } = useParams();
  namespace = namespace?.toLowerCase();
  project = project?.toLowerCase();
```

**Comment by nsheff** [comment]
Might it be better solved at the server-side?

**Comment by nleroy917** [comment]
Could we put it in both places?

**Comment by khoroshevskyi** [comment]
> Might it be better solved at the server-side?

I think it's better solution. We have to use simply lower(str) function before uploading projects to the database

**Comment by nleroy917** [comment]
> I think it's better solution. We have to use simply lower(str) function before uploading projects to the database

Well, we should also think about backwards compatibility if there's already projects in the database with casing issues. That's why on the server is nice too. **However**, whats a clean way to do this across all endpoints? Most, if not all, endpoints bring in `namespace` and I am not too keen on adding `namespace = namespace.lower()` to the top of each and every endpoint.

Is there a clean dependency injection solution for this? Maybe `get_namespace_lower()` ?

**Comment by nleroy917** [comment]
I'm partial to a middleware solution: https://github.com/tiangolo/fastapi/issues/826

```python
class CaseInsensitiveNamesapceMiddleware(BaseHTTPMiddleware):
    """
    This middleware will convert any namespace in the URL to lowercase.
    """

    async def dispatch(self, request: Request, call_next: Send) -> Response:
        # set the environment variables
        namespace = request.path_params.get("namespace")
        if namespace:
            request.path_params["namespace"] = namespace.lower()

        response = await call_next(request)

        return response
```

**Comment by nsheff** [comment]
> 

Don't worry about projects already in the db with casing issues, just delete them :)

**Comment by nleroy917** [comment]
> I'm partial to a middleware solution: [tiangolo/fastapi#826](https://github.com/tiangolo/fastapi/issues/826)
> 
> ```python
> class CaseInsensitiveNamesapceMiddleware(BaseHTTPMiddleware):
>     """
>     This middleware will convert any namespace in the URL to lowercase.
>     """
> 
>     async def dispatch(self, request: Request, call_next: Send) -> Response:
>         # set the environment variables
>         namespace = request.path_params.get("namespace")
>         if namespace:
>             request.path_params["namespace"] = namespace.lower()
> 
>         response = await call_next(request)
> 
>         return response
> ```

Unfortunately, `path_params` are empty since this comes in before the request so this wont work.

**Comment by nleroy917** [comment]
Could be a good use case for their `Annotated` class: https://fastapi.tiangolo.com/tutorial/dependencies/?h=annotated#share-annotated-dependencies

**Comment by nsheff** [comment]
> Unfortunately, `path_params` are empty since this comes in before the request so this wont work.

Isn't there middleware that could come in at the spot that has all the information you need?

**Comment by nleroy917** [comment]
Is this still an issue? Should we open another one about case sensitivity?

**Comment by khoroshevskyi** [comment]
I think this issue is solved in pepdbagent 0.4.0 release. 
![image](https://github.com/pepkit/pephub/assets/41573628/0f860b2b-a085-4fbb-8ebc-97ebc569e002)

**Comment by nleroy917** [comment]
This seems fixed in both `pephub` and in `pepdbagent`. So, we feel this is solved.
